File: 백준/Gold/1477. 휴게소 세우기/휴게소 세우기.py
import sys, math
input = sys.stdin.readline

n, m, L = map(int, input().split())
term_list = []

def get_rest_num(max_term):
    rest_num = 0
    for term in term_list:
        if term <= max_term:
            rest_num += 1
        else:
            rest_num += math.ceil(term / max_term)
        
    return rest_num

def select_term():
    left = 1
    right = L - 1
    total = n + m + 1
    answer = sys.maxsize
    while left <= right:
        mid = (left + right) // 2
        rest_num = get_rest_num(mid)
        if rest_num <= total:
            answer = min(answer, mid)
            right = mid - 1
        else:
            left = mid + 1

    return answer


rest_list = [0] + list(map(int, input().split())) + [L]
rest_list.sort()

for i in range(n+1):
    term_list.append(rest_list[i+1] - rest_list[i])

print(select_term())

File: 백준/Gold/1477. 휴게소 세우기/휴게소 세우기.js
const input = require('fs').readFileSync(process.platform === 'linux' ? '/dev/stdin': __dirname + '/input.txt').toString().split('\n')

const [N, M, L] = input[0].split(' ').map(Number)
const rest = input[1].split(' ').map(Number)
rest.push(0, L)
rest.sort((a, b) => a - b)

const gap = []
for (let i = 0; i < rest.length - 1; i++) {
  gap.push(rest[i + 1] - rest[i])
}
gap.sort((a, b) => a - b)

const answer = set_max_gap()
console.log(answer)

function set_max_gap() {
  let left = 0
  let right = L - 1
  let max_gap = L - 1

  while(left <= right) {
    const gap = Math.floor((left + right) / 2)
    const cnt = build_rest(gap)

    if (cnt <= M) {
      right = gap - 1
      max_gap = Math.min(max_gap, gap)
    } else {
      left = gap + 1
    }
  }

  return max_gap
}

function build_rest(max_gap) {
  let cnt = 0
  gap.forEach((cur) => {
    if (max_gap < cur) {
      cnt += Math.ceil(cur / max_gap) - 1
    }
  })

  return cnt
}